File: langgraph/src/graphs/how-it-works/nodes/read.ts
import { readdirSync, readFileSync, existsSync } from "fs";
import { join, relative } from "path";
import type { State } from "../graph.js";
import type { FileContent } from "../types.js";

// ─── Limits ──────────────────────────────────────────────────────────────────

const MAX_FILE_CHARS = 3_000;
const MAX_TOTAL_CHARS = 28_000;

// ─── Helpers ─────────────────────────────────────────────────────────────────

function readFile(
  absPath: string,
  appPath: string,
  maxChars = MAX_FILE_CHARS
): FileContent | null {
  try {
    const raw = readFileSync(absPath, "utf-8");
    const content =
      raw.length > maxChars ? raw.slice(0, maxChars) + "\n// ... (truncated)" : raw;
    return { relativePath: relative(appPath, absPath), content };
  } catch {
    return null;
  }
}

/** Recursively find page.tsx files up to `maxDepth` directory levels. */
function findPageFiles(dir: string, depth = 0, maxDepth = 3): string[] {
  if (depth > maxDepth) return [];
  const results: string[] = [];
  try {
    for (const entry of readdirSync(dir, { withFileTypes: true })) {
      const full = join(dir, entry.name);
      if (entry.isDirectory()) {
        const skip = ["node_modules", ".next", "how-it-works", "dist", ".git"];
        if (!skip.includes(entry.name)) {
          results.push(...findPageFiles(full, depth + 1, maxDepth));
        }
      } else if (entry.name === "page.tsx" || entry.name === "page.ts") {
        results.push(full);
      }
    }
  } catch {
    // ignore unreadable dirs
  }
  return results;
}

/** Collect TS/TSX files from a flat directory (max 5). */
function readFlatDir(dir: string, max = 5): string[] {
  if (!existsSync(dir)) return [];
  try {
    return readdirSync(dir, { withFileTypes: true })
      .filter((e) => e.isFile() && /\.(ts|tsx)$/.test(e.name))
      .map((e) => join(dir, e.name))
      .slice(0, max);
  } catch {
    return [];
  }
}

// ─── Node ────────────────────────────────────────────────────────────────────

export async function readNode(state: State): Promise<Partial<State>> {
  const app = state.currentApp!;
  const files: FileContent[] = [];
  let totalChars = 0;

  const add = (absPath: string, maxChars = MAX_FILE_CHARS) => {
    if (totalChars >= MAX_TOTAL_CHARS) return;
    const f = readFile(absPath, app.path, maxChars);
    if (f) {
      files.push(f);
      totalChars += f.content.length;
    }
  };

  // 1. package.json — always (framework, deps, scripts)
  add(join(app.path, "package.json"), 2_000);

  // 2. Config files
  for (const name of [
    "next.config.ts",
    "next.config.js",
    "drizzle.config.ts",
    "middleware.ts",
    "middleware.tsx",
    ".env.example",
    "env.example",
  ]) {
    add(join(app.path, name), 1_500);
  }

  // 3. Root layout
  add(join(app.appDir, "layout.tsx"), 2_000);

  // 4. Home page
  const homePage = join(app.appDir, "page.tsx");
  add(homePage);

  // 5. Other pages (up to 10, skip how-it-works)
  const otherPages = findPageFiles(app.appDir)
    .filter((p) => p !== homePage)
    .slice(0, 10);
  for (const p of otherPages) add(p);

  // 6. Lib / utils / server directories (up to 5 files each, 2 dirs max)
  for (const dir of ["lib", "src/lib", "utils", "src/utils", "server", "src/server"]) {
    for (const p of readFlatDir(join(app.path, dir), 5)) add(p);
    if (files.length > 30) break;
  }

  // 7. API routes (up to 5)
  const apiDir = join(app.appDir, "api");
  for (const p of findPageFiles(apiDir).slice(0, 5)) add(p);

  // 8. Schema / DB files
  for (const dir of ["src/db", "db", "schema"]) {
    for (const p of readFlatDir(join(app.path, dir), 4)) add(p);
  }

  console.log(
    `  📁  Read ${files.length} files  (${Math.round(totalChars / 1_000)}k chars)`
  );
  if (state.verbose) {
    for (const f of files) console.log(`       ${f.relativePath}`);
  }

  return { currentFiles: files };
}

File: langgraph/src/how_it_works/nodes/read.py
"""Read node — extracts source files from the current app."""

from __future__ import annotations

import re
from pathlib import Path
from typing import Any

from how_it_works.models import FileContent

MAX_FILE_CHARS = 3_000
MAX_TOTAL_CHARS = 48_000


def _read_file(
    abs_path: Path, app_path: Path, max_chars: int = MAX_FILE_CHARS
) -> FileContent | None:
    try:
        raw = abs_path.read_text(encoding="utf-8")
        content = (
            raw[:max_chars] + "\n// ... (truncated)" if len(raw) > max_chars else raw
        )
        return FileContent(
            relative_path=str(abs_path.relative_to(app_path)), content=content
        )
    except (OSError, UnicodeDecodeError):
        return None


def _find_page_files(directory: Path, depth: int = 0, max_depth: int = 3) -> list[Path]:
    if depth > max_depth:
        return []
    results: list[Path] = []
    skip = {"node_modules", ".next", "how-it-works", "dist", ".git"}
    try:
        for entry in sorted(directory.iterdir()):
            if entry.is_dir() and entry.name not in skip:
                results.extend(_find_page_files(entry, depth + 1, max_depth))
            elif entry.name in ("page.tsx", "page.ts"):
                results.append(entry)
    except OSError:
        pass
    return results


def _read_flat_dir(directory: Path, max_files: int = 5, ext_pattern: str = r"\.(ts|tsx)$") -> list[Path]:
    if not directory.is_dir():
        return []
    try:
        return [
            f
            for f in sorted(directory.iterdir())
            if f.is_file() and re.search(ext_pattern, f.name)
        ][:max_files]
    except OSError:
        return []


async def read_node(state: dict[str, Any]) -> dict[str, Any]:
    app = state["current_app"]
    app_path = Path(app.path)
    app_dir = Path(app.app_dir)

    files: list[FileContent] = []
    total_chars = 0

    def add(abs_path: Path, max_chars: int = MAX_FILE_CHARS) -> None:
        nonlocal total_chars
        if total_chars >= MAX_TOTAL_CHARS:
            return
        f = _read_file(abs_path, app_path, max_chars)
        if f:
            files.append(f)
            total_chars += len(f.content)

    # 1. package.json
    add(app_path / "package.json", 2_000)

    # 2. Config files
    for name in [
        "next.config.ts",
        "next.config.js",
        "drizzle.config.ts",
        "middleware.ts",
        "middleware.tsx",
        ".env.example",
        "env.example",
    ]:
        add(app_path / name, 1_500)

    # 3. Root layout
    add(app_dir / "layout.tsx", 2_000)

    # 4. Home page
    home_page = app_dir / "page.tsx"
    add(home_page)

    # 5. Other pages (up to 10)
    other_pages = [p for p in _find_page_files(app_dir) if p != home_page][:10]
    for p in other_pages:
        add(p)

    # 6. Lib / utils / server dirs
    for dir_name in ["lib", "src/lib", "utils", "src/utils", "server", "src/server"]:
        for p in _read_flat_dir(app_path / dir_name, 5):
            add(p, 4_000)
        if len(files) > 30:
            break

    # 7. API routes
    api_dir = app_dir / "api"
    for p in _find_page_files(api_dir)[:5]:
        add(p)

    # 8. Schema / DB files
    for dir_name in ["src/db", "db", "schema"]:
        for p in _read_flat_dir(app_path / dir_name, 4):
            add(p)

    # 9. Evals (test/eval files)
    for p in _read_flat_dir(app_path / "evals", 10, r"\.(py|yaml|yml|ts)$"):
        add(p)

    # 10. App-local LangGraph pipelines
    for p in _read_flat_dir(app_path / "langgraph", 8, r"\.py$"):
        add(p)
    for p in _read_flat_dir(app_path / "langgraph" / "src", 8, r"\.py$"):
        add(p)

    # 11. Prompts
    for p in _read_flat_dir(app_path / "prompts", 5, r"\.(txt|md|yaml|ts)$"):
        add(p)

    # 12. Root configs
    for name in ["promptfooconfig.yaml", "pytest.ini", "pyproject.toml"]:
        add(app_path / name, 1_500)

    print(f"  📁  Read {len(files)} files  ({round(total_chars / 1_000)}k chars)")
    if state.get("verbose"):
        for f in files:
            print(f"       {f.relative_path}")

    return {"current_files": files}
